cairo/tests/ethereum/cancun/test_fork.py

```
@@ -98,3 +101,24 @@ def test_make_receipt(
         assert make_receipt(tx, error, cumulative_gas_used, logs) == cairo_run(
             "make_receipt", tx, error, cumulative_gas_used, logs
         )
+
+    @given(env=..., tx=...)
+    def test_process_transaction(self, cairo_run, env: Environment, tx: Transaction):
+
+        try:
+            gas_used_cairo, logs_cairo, error_cairo = cairo_run(
+                "process_transaction", env, tx
+            )
+        except Exception as cairo_exception:
+            try:
+                process_transaction(env, tx)
+            except Exception as python_exception:
+                error = re.search(r"Error message: (.*)", str(cairo_exception))
+                error_type = error.group(1) if error else type(cairo_exception).__name__
+                assert type(python_exception).__name__ == error_type
```

what is this?

---

We must execute the cairo code before the python one as we cannot do copy of the input (due to frozen properties) so the exact error type raise is not known.

The thing is: cairo will raise an AssertionError when there is an ASSERT_EQ. In this function we use those ASSERT_EQ to raise a specific error and we do not know the type of the error raise unless we do this.
To retrieve the error, we need to compare type raised by the python code. 
So this hack, open to better ideas
 
```
def map_to_python_exception(e: Exception):
    error_str = str(e)

    # Throw a specialized python exception from the error message, if possible
    error = re.search(r"Error message: (.*)", error_str)
    error_type = error.group(1) if error else error_str
    exception_class = __builtins__.get(error_type)
    if isinstance(exception_class, type) and issubclass(exception_class, Exception):
        raise exception_class() from e

    # Fallback to generic exception
    if "An ASSERT_EQ instruction failed" in error_str:
        raise AssertionError(e) from e
    raise Exception(error_str) from e
```

---

I'm not sure I get it 🤔 

We raise ` AssertionError when there is an ASSERT_EQ` if there is no error message the assertion is wrapped with. Otherwise we are supposed to raise with the correct error type, which is taken from the with_attr statement

---

For example this one: 
```
with_attr error_message("InvalidBlock") {
        assert is_valid.value = TRUE;
    }
```

This will raise with an AssertionError because "An ASSERT_EQ instruction failed" is in the error.
But it will also contain the error_message InvalidBlock

---

I think InvalidBlock is a legacy piece of code since it's supposed to be an EthereumException

The rule for now is:
- if the error raised is a python std lib error, you can do 
`with_attr error_message("ExactNameOfTheError") { assert is_valid.value = TRUE; }` and then the runner will convert it to the right python error
- if the error raised is an ethereumexception, apply the same rules as for the other ethereum exceptions

---

InvalidBlock is raised in EELS and catch anywhere. It is even raised by the state_transition function
https://github.com/ethereum/execution-specs/blob/d9febcb7332d62194968dcdb7775adf497c93626/src/ethereum/cancun/fork.py#L177

---

> This will raise with an AssertionError because "An ASSERT_EQ instruction failed" is in the error.

Surprising, this should be handled by the runner. let me check

---

@obatirou pushed a fix. 
@eikix this will also be necessary for your PR

---

@obatirou pushed a fix. 
@eikix this will also be necessary for your PR

---

thank you 
Should we let as is then ?

---

might have to limit the number of examples also as it is very long even with custom strategy

---

> @obatirou pushed a fix. @Eikix this will also be necessary for your PR

ok i'll rebase and use this to modify the signature of recover_sender to raise instead of return the exception